agent/consul/client_test.go

```diff
@@ -896,33 +895,53 @@ func TestClient_RPC_Timeout(t *testing.T) {
 	})
 
 	// waiter will sleep for 101ms which is 1ms more than the DefaultQueryTime
-	require.NoError(t, s1.RegisterEndpoint("Wait", &waiter{duration: 101 * time.Millisecond}))
+	require.NoError(t, s1.RegisterEndpoint("Long", &waiter{duration: 101 * time.Millisecond}))
+	require.NoError(t, s1.RegisterEndpoint("Short", &waiter{duration: 5 * time.Millisecond}))
 
-	// Requests with QueryOptions have a default timeout of RPCHoldTimeout (10ms)
-	// so we expect the RPC call to timeout.
-	var out struct{}
-	err := c1.RPC("Wait.Wait", &structs.NodeSpecificRequest{}, &out)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "rpc error making call: i/o deadline reached")
-
-	// Blocking requests have a longer timeout (100ms) so this should pass since we
-	// add the maximum jitter which should be 16ms
-	out = struct{}{}
-	err = c1.RPC("Wait.Wait", &structs.NodeSpecificRequest{
-		QueryOptions: structs.QueryOptions{
-			MinQueryIndex: 1,
-		},
-	}, &out)
-	require.NoError(t, err)
+	t.Run("non-blocking query times out after RPCReadTimeout", func(t *testing.T) {
+		// Requests with QueryOptions have a default timeout of RPCReadTimeout (10ms)
+		// so we expect the RPC call to timeout.
+		var out struct{}
+		err := c1.RPC("Long.Wait", &structs.NodeSpecificRequest{}, &out)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "rpc error making call: i/o deadline reached")
+	})
 
-	// We pass in a custom MaxQueryTime (20ms) through QueryOptions which should fail
-	out = struct{}{}
-	err = c1.RPC("Wait.Wait", &structs.NodeSpecificRequest{
-		QueryOptions: structs.QueryOptions{
-			MinQueryIndex: 1,
-			MaxQueryTime:  20 * time.Millisecond,
-		},
-	}, &out)
-	require.Error(t, err)
-	require.Contains(t, err.Error(), "rpc error making call: i/o deadline reached")
+	t.Run("non-blocking query succeeds", func(t *testing.T) {
+		var out struct{}
+		require.NoError(t, c1.RPC("Short.Wait", &structs.NodeSpecificRequest{}, &out))
+	})
+
+	t.Run("check that deadline does not persist across calls", func(t *testing.T) {
+		var out struct{}
+		err := c1.RPC("Long.Wait", &structs.NodeSpecificRequest{}, &out)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "rpc error making call: i/o deadline reached")
+		// We use structs.KVSRequest, which does not implement pool.BlockableQuery
+		// and should have no timeouts defined.
+		require.NoError(t, c1.RPC("Long.Wait", &structs.KVSRequest{}, &out))
+	})
+
+	t.Run("blocking query succeeds", func(t *testing.T) {
+		// Blocking requests have a longer timeout (100ms) so this should pass since we
+		// add the maximum jitter which should be 16ms
```

Comment:
If these tests depend on timing differences of 16ms then they'll probably end up being pretty flakey in CI right? Or did I misunderstand the comment?

Reply:
Blocking query default timeout (in the test setup): 100ms
`Long.Wait` returns after: 101ms
The max jitter is added when calculating blocking query timeouts: 100ms + 100/16 = 116ms

So theoretically if max jitter is ever applied by the server and it returns in 117ms then the test will flake. I will make the `Long.Wait` exactly 100ms

Reply:
Side comment to consider in "the future":

Do we have a lot of timing dependent (possibly flaky) tests? If so, it might be worthwhile at some point to develop a mockable interface passed in via dependency injection so we can use real timing capabilities in the real code, but mocked versions (that can't flake) in the tests.

Reply:
Agreed! We actually have an established mock in a few places like https://github.com/hashicorp/consul/blob/80e51ff907b2b68ac42e0d29773e480d07e59bd1/agent/grpc-external/services/peerstream/testing.go#L136

Reply:
In the past timing-dependent tests have needed error margins in the order of 500ms or more to not be very flaky in CI. 

What happens if CI CPU is starved and so the goroutine handling the RPC isn't scheduled for 200ms and so doesn't "notice" the timeout, that would flake right?

Usually I'd treat tests like this that use actual non-deterministic goroutines and wall clock time as a very very rough smoke test that the timeout does anything at all rather than trying to assert anything very specific about exactly how long the timeout was. i.e. make it so that if you comment out the timeout code it always fails but maybe after like several _seconds_ then you can use much larger margins.

All that said, I've been much further from out test suite than you for a long time @kisunji so if you have good reason to think this will be robust I defer to your greater knowledge! I just know that any sleep-style test with timing tolerances lower than multiple hundreds of milliseconds used to be a non-starter in CI 😆 !